### ISSUE TITLE
"Open With" support

### ISSUE THREAD

**Issue by CrossR** [extr0py/oni#598]
The last big thing that is preventing me from using Oni full time is proper integrations into my OS as an editor.

For me this means:
* Oni registered as a file editor
  * This seems to already be the case.
* "Open With" support.
  * Ie, I can double click a text file and it opens in Oni.
* "Recent File" support.
  * This looks to be implemented, since I've got it in my Windows 10 start menu, but clicking the files doesn't actually open them. I expect this is due to the above bullet point.

Really, its just the "Open With" support. With that, I can drop using Notepad++ on Windows 10 and start using Oni for my poking around files and such.

Is this a Windows specific issue, or just that its not been implemented? I can have a look into how you are supposed to do this with Electron, but I'm not the most knowledgeable with it.

**Comment by extr0py** [comment]
Hi @CrossR ! Thanks for logging this issue. These sorts of blockers are really helpful to call out, as I'm sure this is a blocker for others as well.

On Windows, the 'Open With' functionality is controlled by a registry key. So we'd need to update our installer to build up those associations. There's a stackoverflow thread here:
https://stackoverflow.com/questions/26536030/inno-setup-file-association

And some more information about the tooling used to create our setup program (InnoSetup): http://www.jrsoftware.org/ishelp/

On OSX, it's controlled by a plist file - this has some more info:
https://stackoverflow.com/questions/28222787/mac-os-x-file-association-works-but-file-icon-not-changed

> With that, I can drop using Notepad++ on Windows 10 and start using Oni for my poking around files and such.

That would be awesome! 😄

**Comment by CrossR** [comment]
Looks like getting the registry keys added isn't so bad, its just a few lines in the setup.template.iss file.
What is awkward is that you need to duplicate those lines for each and every extension you want to support...

If you look at the VSCode .iss file here, you can see its just massive : https://github.com/Microsoft/vscode/blob/master/build/win32/code.iss

Since its so easy to add yourself (Right Click -> Open With -> Look for another app on PC -> Point at oni.exe), I think what makes most sense is just adding a few for now, and then add to it as language servers are added?

That all said, even after Oni is registered as the editor, it doesn't open any files when I double click a file, even when I've said to open with Oni.

I'm less sure how I should go about doing this linking, it looks like on Windows at least the file path you want to load is stored in argv[1], so I need to grab that somewhere and then open the file it says. Just can't work out where the code actually enter when I use the open with dialog.

**Comment by extr0py** [comment]
> Looks like getting the registry keys added isn't so bad, its just a few lines in the setup.template.iss file.
What is awkward is that you need to duplicate those lines for each and every extension you want to support...

Cool, nice find! It looks like most of the entries are duplicates, except for the file extensions.... Since we already use a javascript script to generate our `iss` file, we could probably reduce that duplication and just have like a javascript array of supported extensions.

> That all said, even after Oni is registered as the editor, it doesn't open any files when I double click a file, even when I've said to open with Oni.

I just tried this too. I think as you mentioned this is due to some differences in how `process.argv` works in both scenarios.

When starting oni from the command-line, from `npm link`, `require("electron").remote.process.argv` returns `["C:\oni\node_modules\electron\dist\electron.exe", "C:\oni\main.js", "TODO.md"]` (__3__ arguments), whereas if it is from the executable, it looks like: `["C:\Program Files (x86)\Oni\Oni.exe", "TODO.md"]` (__2__ arguments)

So I believe we might not be handling the arguments correctly... we might be assuming there are always _two_ arguments prior to the file, but in this case there is actually only one.

**Comment by CrossR** [comment]
> Cool, nice find! It looks like most of the entries are duplicates, except for the file extensions.... Since we already use a javascript script to generate our iss file, we could probably reduce that duplication and just have like a javascript array of supported extensions.

That made a lot more sense so I've swapped to that instead. From checking out the VSCode implementation I also realised how easy it was to add Oni to the Windows Path (#344) so I've done that too.

> So I believe we might not be handling the arguments correctly... we might be assuming there are always two arguments prior to the file, but in this case there is actually only one.

Yep, thats the case, in main.js we do `createWindow(process.argv.slice(2), process.cwd())`. I'm not fully sure how to go about sorting that though.

There is two cases when there is two arguments:
1. When using open with you have the .exe and the file.
2. If you call Oni from the command line with no file to open, you have electron.exe and main.js

So we can't just check if there is two arguments and use the second, since the second could either be a file, or main.js. The easy approach is just check if the first argument is oni or electron, but not sure if thats a smart thing to do either.
